### PR TITLE
[pt/sft] Feature channel loss

### PR DESCRIPTION
**Pull request by kevssim**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [x] Document Updates
- [ ] More Models or Datasets Support

# PR information

Monitoring the model's performance on different tasks(channels).

### CODE REVIEW

File: swift/llm/dataset/preprocessor/core.py [hunk 1]
```diff
@@ -323,6 +328,11 @@ def __call__(
         if isinstance(dataset_mapped, HfDataset) and len(dataset) != len(dataset_mapped):
             logger.info(
                 f'Dataset filtered, origin length: {len(dataset)}, filtered dataset length: {len(dataset_mapped)}')
+        if channel:
```
Comment (Jintao-Huang) [inline swift/llm/dataset/preprocessor/core.py]: 这里是为什么呢

File: swift/trainers/trainers.py [hunk 2]
```diff
@@ -88,6 +88,8 @@ def __init__(self, *args, **kwargs):
             self.infer_engine = PtEngine.from_model_template(
                 self.model, self.template, max_batch_size=self.args.per_device_eval_batch_size)
         self.jsonl_writer = JsonlWriter(os.path.join(self.args.output_dir, 'predict.jsonl'))
+        self.channel_cid = None
```
Comment (Jintao-Huang) [inline swift/trainers/trainers.py]: 这个写个注释吧

channel_cid和cid_channel

File: swift/llm/dataset/preprocessor/core.py (hunk 1 above)
Comment (kevssim) [reply swift/llm/dataset/preprocessor/core.py]: 这里是为了读取数据的‘channel’字段，把它放到inputs中。
数据示例：{
   "channel": "chat",
   "messages": [
    {"role": "system", "content": "You are a helpful assistant"},
    {"role": "user", "content": "What color do you like?"},
    {"role": "assistant", "content": "I like blue."}]
}

File: swift/trainers/trainers.py (hunk 2 above)
Comment (kevssim) [reply swift/trainers/trainers.py]: 已修改为通过参数方式传入，不需要channel_cid和cid_channel了

File: docs/source/Instruction/命令行参数.md [hunk 3]
```diff
@@ -351,6 +351,7 @@ Vera使用`target_modules`, `target_regex`, `modules_to_save`三个参数.
 - check_model: 检查本地模型文件有损坏或修改并给出提示，默认为True。如果是断网环境，请设置为False。
 - 🔥create_checkpoint_symlink: 额外创建checkpoint软链接，方便书写自动化训练脚本。best_model和last_model的软链接路径分别为f'{output_dir}/best'和f'{output_dir}/last'。
 - loss_type: loss类型。默认为None，使用模型自带损失函数。
+- channel_list : 数据集包含的channel列表。默认为None。结合`--loss_type channel_loss`使用，可参考[这里](https://github.com/modelscope/ms-swift/blob/main/examples/train/plugins/channel_loss.sh)。
```
Comment (Jintao-Huang) [inline docs/source/Instruction/命令行参数.md]: 名字请改成channels

这有可能在训练的时候自动记录channel的实现么

File: docs/source_en/Instruction/Command-line-parameters.md [hunk 4]
```diff
@@ -360,6 +360,7 @@ Training arguments include the [base arguments](#base-arguments), [Seq2SeqTraine
 - check_model: Check local model files for corruption or modification and give a prompt, default is True. If in an offline environment, please set to False.
 - 🔥create_checkpoint_symlink: Creates additional checkpoint symlinks to facilitate writing automated training scripts. The symlink paths for `best_model` and `last_model` are `f'{output_dir}/best'` and `f'{output_dir}/last'` respectively.
 - loss_type: Type of loss. Defaults to None, which uses the model's built-in loss function.
+- channel_list：List of channels included in the dataset. Defaults to None. Used in conjunction with `--loss_type channel_loss`. Refer to [this example](https://github.com/modelscope/ms-swift/blob/main/examples/train/plugins/channel_loss.sh) for more details.
```
Comment (Jintao-Huang) [inline docs/source_en/Instruction/Command-line-parameters.md]: +1

File: swift/llm/train/sft.py [hunk 5]
```diff
@@ -116,6 +116,7 @@ def run(self):
             train_dataset=train_dataset,
             eval_dataset=val_dataset,
             callbacks=self.callbacks,
+            channel_list=self.args.channel_list,
```
Comment (Jintao-Huang) [inline swift/llm/train/sft.py]: 这个写在TrainArgumentsMixin吧，不在trainer的init方法传入了

File: swift/llm/dataset/preprocessor/core.py (hunk 1 above)
Comment (Jintao-Huang) [reply swift/llm/dataset/preprocessor/core.py]: 这里的map应该是不需要的。dataset_mapped中应该已经包含channel了

Comment (kevssim) [reply swift/llm/dataset/preprocessor/core.py]: image
这一步会把channel删掉，是否改成把channel从remove_columns中移除？

File: docs/source/Instruction/命令行参数.md (hunk 3 above)
Comment (kevssim) [reply docs/source/Instruction/命令行参数.md]: 已处理，见https://github.com/modelscope/ms-swift/pull/4405/commits/48b4aecb6684ac938daf69a40afea7f5cfb4f7d0

没有在训练的时候自动记录的原因是：多卡情况下，需要保持channels一致(否则可能死锁)，自动记录可能会导致不一致，需要通信来同步，增加通信开销

File: swift/llm/train/sft.py (hunk 5 above)
Comment (kevssim) [reply swift/llm/train/sft.py]: 已处理，见https://github.com/modelscope/ms-swift/pull/4405/commits/48b4aecb6684ac938daf69a40afea7f5cfb4f7d0

File: swift/llm/dataset/preprocessor/core.py (hunk 1 above)
Comment (kevssim) [reply swift/llm/dataset/preprocessor/core.py]: 已处理，见https://github.com/modelscope/ms-swift/pull/4405/commits/c12f56666b46597355f149febb8e152f349806e3

**Review by Jintao-Huang — APPROVED** [review]

**Comment by Jintao-Huang** [comment]
还有更新不，打算merge了

**Comment by kevssim** [comment]
没有更新了，merge吧，谢谢
